File: 5_kyu/Mean_without_outliers.r
# Calculates arithmetic mean value of the samples without outliers.
# Outliers are samples with distance to mean > cutoff*population standard deviation.
clean_mean <- function(sam, cutoff){
  n <- 1L
  subset <- sam
  m <- mean(subset)
  repeat {
    s <- sd.population(subset)
    subset <- subset[abs(subset-m) <= cutoff * s]
    m_next <- mean(subset)
    if( abs(m-m_next) < 1e-5*max(1,abs(m)) | n > 10L ) break
    n <- n + 1L
    m <- m_next
  }
  round(m,digits=2)
}

# Standard deviation of the population
sd.population = function(x) sd(x)*sqrt((length(x)-1)/length(x))

File: 5_kyu/Mean_without_outliers.py
import math
import statistics as stat
import typing as ty

# Returns mean value of the sample.
# Calculate mean value only with samples around cutoff*stdev around the value.
def clean_mean(sample: ty.List[float], cutoff: float) -> float:
    s = sample
    m = stat.mean(s)
    # Limit the maximal number of iterations to prevent oscillations
    for i in range(10*len(sample)):
        limit = cutoff * math.sqrt(stat.pvariance(s,mu=m))
        ll, ul = m-limit, m+limit
        s = [x for x in sample if ll < x < ul]
        old_m = m
        m = stat.mean(s)
        if abs(old_m-m) < 1/100: break
    return round(m,2)
